contents/blog/async-migrations.md

```diff
@@ -0,0 +1,185 @@
+
+### 3. It should be easy to work with, both for developers and users**
+
+Lastly, async migrations should be easy to launch and manage for users, as well as easy to write and build upon for developers.
+
+To cover the latter part, we implemented the aforementioned guarantees and checks, as well as made the migration definitions closely resemble other migration systems we already have in place.
+
+As for users, we built a fully-fledged UI for instance admins, where they can read about the migration, update settings, trigger a run, pause or resume a migration, force a rollback, monitor status and progress, as well as get a list of errors. 
+
+We also store the ID of the query that is currently running on the database and the ID of the Celery task handling the migration, so more advanced users can go into the services themselves for additional monitoring.
+
+Ultimately, running a migration takes only the click of a button, but all the tools are in place for debugging issues if they come about. 
+
+It's also worth noting that we implemented an API for migrations to check if they should even run at all on a given instance. This is particularly useful for preventing useless runs on fresh deploys. A user that deployed PostHog after the migration was introduced shouldn't need to run a migration to get from an old schema to the new schema, given they already have the new schema in place. In these cases, we auto complete the migration and unblock upgrades without any user input at all.
+
+## A closer look
+
+### Architecture
+
+Async migrations are triggered from the UI and create a Celery task that handles its execution in a background worker. 
+
+Metadata about the migration is stored in Postgres, including things such as status (completed, errored, rolled back, etc), progress, start and finish times, and other useful information.
+
+![Async migrations Architecture](../images/blog/async-migrations-architecture.png)
+
+### Structure
+
+An async migration can contain the following components:
+
+* `description`: A description of the migration.
+* `operations`: A list of operations to run in sequence, each with a designated rollback. Operations can be plain SQL or an arbitrary function.
+* `posthog_min_version`: The minimum version of PostHog required for running the migration.
+* `posthog_max_version`: The last PostHog version this migration can be run on.
+* `is_required`: A check that determines if this migration needs to be run on the instance. Usually verifies the current schema of a table to determine if it needs to be migrated.
+* `precheck`: A check that runs before the migration starts to determine if it is safe to do so. It returns a boolean and a message, so it can surface helpful errors to the user. 
+* `healthcheck`: A check that runs both before the migration starts and at an interval while it is running. If it fails, it will kill the migration. Also returns a message that will be shown to the user.
```

Comment:
maybe add sth like: and roll it back if automatic rollbacks are enabled (the default). This will ensure we're not stuck in a bad state.